### ISSUE TITLE
Updating git secrets

### ISSUE THREAD

**Issue by Olshansk** [loderunner/scrt#577]
**Question**: How does one update the secrets on github?

1. I have the secrets set up locally
2. I have initialised a repote repo
3. Looking at [reference/storage/git](https://scrt.run/reference/storage/git.html), it's unclear how one would push changes?

**Comment by loderunner** [comment]
There might be a confusion here. Local and git backends are not supposed to be used on the same store. The "storage backends" are really just the way scrt writes and persists your secrets.

With the git storage backend, the repository is meant to be used as a secret storage system as a whole. If you wish to maintain a secrets file alongside other code in the repository, just use the local storage backend, and commit changes to the store file like regular code. This is how I manage secrets for the scrt project itself.

---

That being said, here is an example of using git as your storage backend.

Let's assume your secrets repo has a remote at `https://github.com/Olshansk/secrets.git` and the secrets file `store.scrt` is located at the root of the repo.

`.scrt.yml`

```yaml
storage: git
git:
  url: https://github.com/Olshansk/secrets.git
  path: store.scrt
```

You can then start using the git remote as your secrets storage:

```shell
git set hello world
```

This will:

* clone your repo from the remote
* checkout the latest revision on main
* set the value `world` for `hello` in the `store.scrt` file
* commit the changes to the file
* push the changes to the remote

Does that solve your case?

**Comment by Olshansk** [comment]
@loderunner Appreciate the response, but I still have a few questions.

For context, I have already initialized a local repo using `scrt init ...` and a remote repo at `https://github.com/Olshansk/secrets.git` using `scrt init --storage=git`.

I use the local one regularly to set and retrieve secrets (works flawlessly), but the fact that the remo repo is not getting updated means that's not the default config.

---

> just use the local storage backend, and commit changes to the store file like regular code.

I realize this is an option and am just considering doing this myself.

---
> .scrt.yml

Where is this yaml file used? Couldn't find documentation for it.

---

> This will:
> 
> clone your repo from the remote
> checkout the latest revision on main
> set the value world for hello in the store.scrt file
> commit the changes to the file
> push the changes to the remote

What happens if git is my default backend but I want to add new secrets while I'm offline?

**Comment by loderunner** [comment]
> > .scrt.yml
>
> Where is this yaml file used? Couldn't find documentation for it.

The usage of a yaml file for configuration is described in the documentation, [in the "Configuration" page](https://scrt.run/guide/configuration.html#configuration-file). Perhaps I should link to this page from the "Getting Started" page?

----

> What happens if git is my default backend but I want to add new secrets while I'm offline?

The "git" backend is actually a "git remote" backend. Reading & writing secrets depends on being able to reach the remote. No offline operations on secrets are possible.

The idea is to rely on git's capabilities for distributed persistence & versioning to store secrets remotely. The git backend is actually meant to be understood as "use a git remote to store your secrets", not "store your secrets inside a git repo". The local backend is much more adequate to this case.

I understand how this can be confusing, I'll see how I can make this clear in the documentation.

**Comment by Olshansk** [comment]
> The usage of a yaml file for configuration is described in the documentation, [in the "Configuration" page](https://scrt.run/guide/configuration.html#configuration-file). Perhaps I should link to this page from the "Getting Started" page?

I think mentioning that the `Configuration` section on the `Getting Started` page should mention "read heard for more details on config files"

---

Yea, this situation is definitely confusing. My personal dream is:

- Local scrt: `git commit`, `git reset`, `git ammend`, etc...
- Remote scrt: `git push`, `git pull`, etc...

This tool is still awesome so I think I'll amend my personal workflow to periodically push my `~/.store.scrt` to my git repo.

**Comment by loderunner** [comment]
I understand the confusion, perhaps I should also document the term "backend" better.

The bottom line is "you can't mix backends". Backends are really just the mechanism through which scrt stores and retrieves your secrets files. They handle transport and persistence. Mixing local git operations with remote git operations would result in an inconsistent API.

Another way to think of the git backend is as a "git remote" backend. It's not meant to replace git operations, it's meant to rely on git remote operations to store secrets.

Yet another way to think of the git backend is as a bit of a hack. I first built scrt for local usage, and then thought: "With git and GitHub, I can get cloud storage, versioning, multi-user, and conflict detection for free!" So, as you point out, the use cases don't match 100%, but at this point, it's the best I could build without writing my own scrt cloud backend.

---

All this being said, there may be a workaround for your case. The git backend relies on the `--git-url` config to know where to locate the remote. You can leverage what a "remote URL" means to git to choose between local and remote operations.

```shell
# Work remotely
export SCRT_GIT_URL=https://github.com/Olshansk/secrets.git

# Work locally
export SCRT_GIT_URL=/home/olshansk/path/to/secrets/repo
```

(or use `--git-url` in the command line, or `git.url` in the config file)

Well... it _should_ work, but trying this out led me to a bug in scrt. I'll get right to fixing it and let you know when it's done.

**Comment by loderunner** [comment]
```
$ scrt init --git-url=/tmp/scrt-git-local --git-path=store.scrt
Error: could not save data to store: command error on refs/heads/main: branch is currently checked out
```

**Comment by Olshansk** [comment]
Will definitely try it out once it's fixed!

We can document this workflow afterwards and hopefully it won't be only for me.

**Comment by loderunner** [comment]
I may have found your workaround.

In order to be used as remote, a git repo should be [bare](https://git-scm.com/book/en/v2/Git-on-the-Server-Getting-Git-on-a-Server), i.e. not have a working directory. This means that, by default, git will not let you push to a locally checked out repo, as remote.

This is to avoid letting HEAD go out of sync with the working tree.

You can, however configure your local git repo to work as remote and accept pushes to the working tree on the current branch. From your local repo:

```shell
git config receive.denyCurrentBranch updateInstead
```

This will let your local non-bare repo accept pushes as updates to the working tree ([docs](https://git-scm.com/docs/git-config#Documentation/git-config.txt-receivedenyCurrentBranch)).

Since these are details of git, and not of scrt, I don't think this fully qualifies as a bug, warranting a change in the code. And since this is all based around hacking your remote URL to point to a local directory, I don't think this deserves to be described up front in the documentation. I have been thinking of adding a "Cookbook" to the docs, though, to describe different workflows. If I get around to it, I'll be sure to talk about this.

If you try this hack out, let me know if it fixes your use case.

**Comment by Olshansk** [comment]
@loderunner I appreciate all the help and support. Ultimately, I've decided to update the path to `SCRT_LOCAL_PATH` and store it in my local configs directly, which I track in a remote GitHub repo as well.

Since you mentioned all the "scrt<-> github-related" integrations are _hacks_, and since I already have a `configs` repo for myself, I figured that storing the `.store.scrt` file alongside all by `.rc` files is the path of least resistance.
